### ISSUE TITLE
Captcha always submitted twice, causing validation error

### ISSUE THREAD

**Issue by sfmskywalker** [OrchardCMS/Orchard#7747]
**TLDR**
When you try to submit a comment with the **ReCaptchaPart** attached, the ReCaptcha driver's `UpdateEditor` method is invoked twice. This causes the captcha to be submitted twice to Google. Causes problems.

**Repro Steps**

1. Setup a new Orchard site (develop branch)
2. Enable the Anti-Spam feature
3. Attach the ReCaptchaPart to the Comment content type
4. Configure the ReCaptcha settings with a valid public and private key (generate them [here](http://www.google.com/recaptcha/admin))
5. Create a dummy Blog and Blog Post so you can test submitting a comment
6. Submit a comment
7. Observe the validation error:

![image](https://user-images.githubusercontent.com/938393/27517005-88b4a0de-59c4-11e7-85bc-31566cdab595.png)

**The Issue**
The core issue is that the captcha is submitted twice. The first time the response is succesful, but the second time, Google says (rightfully so) that the captcha was already submitted. This causes a validation error.
The reason the `UpdateEditor` method is invoked twice is because the `CommentController` invokes `UpdateEditor` on the Comment content item twice.

**Possible Remedies**
There are a number of ways to handle this.

1. The root cause is the fact that the `CommentController` invokes `UpdateEditor` on the Comment twice. There's probably a good reason for this, so we may not want to change this behavior. At the same time, the need for having to call `UpdateEditor` twice might be a symptom of an underlying problem with the way `ContentManager.Create` works on a content item on which `UpdateEditor` was already called.
2. If we can't update the `CommentController` to only invoke `UpdateEditor` once, we can update the `ReCaptchaPartDriver` to "remember" that it already submitted the captcha.
3. Alternatively, we can simply check the return code from Google, and ignore the "duplicate" error code by not adding a validation error.

PR: #7748

**Comment by sfmskywalker** [comment]
I'm gonna try and see why the UpdateEditor method is called twice from the CommentController first. Ideally we can just remove the second call, or come up with a better fix than having to call UpdateEditor after ContentManager.Create.

**Comment by sfmskywalker** [comment]
Everything seems to work just fine if I remove the second call to UpdateEditor, so I'm going to submit a PR. If anyone remembers why we do need to call UpdateEditor twice, I'll send another PR with a comment explaining the second call, in addition to proposed solution 2 or 3.

**Comment by sfmskywalker** [comment]
PR: #7748

**Comment by sfmskywalker** [comment]
I noticed that this bug does not exist in 1.10.x (UpdateEditor is only called once). When checking the history of the CommentController.cs file, I have not seen any version that calls UpdateEditor twice, which means the second call must have been introduced in the dev branch.

**Comment by sfmskywalker** [comment]
Found it. The second call was indeed added to the dev branch as part of commit b7414e6f  to fix issue #6733 (Comments and Audit Trail issue).

**Comment by sfmskywalker** [comment]
Looking at the comments of issue #6733, the following explains the addition of the second call to UpdateEditor:

> To fix this, Orchard.Blogs.BlogPostAdminController.CreatePOST(int blogId, bool publish) is a good example on how to implement content item creation. CommentController.Create(...) should ContentManager.Create the CommentPart first, only then use ContentManager.UpdateEditor(...) to update part and validate updated part. After this:
> 
> If everything is correct, all DB changes will be saved upon request finishes. -OR-
> If something goes wrong, say ModelState.IsValid == false, cancel the DB transaction via Services.TransactionManager.Cancel().
> 

It seems to me we should still remove the second call, but change the order a bit such that ContentManager.Create is called first, then do the UpdateEditor, and if model validation fails, roll back the transaction.

I'll try this out and update the PR when this works.

**Comment by sfmskywalker** [comment]
I updated the PR by moving the call to `ContentManager.Create` up so that it executes before the call to `ContentManager.UpdateEditor` to prevent the issue with Audit Trail as described in #6733.

The code in `CommentController.Create` already handled rolling back the transaction in case of a model validation error, which matches the pattern used in `BlogPostAdminController` (which is referenced as an example in #6733).

**Comment by sebastienros** [comment]
Calling ContentManager.Create first will create some transaction log which is an issue. It was done in this order on purpose.

**Comment by MatteoPiovanelli** [comment]
Hello. I am resurrecting this because we are seeing a similar issue (recaptcha posted/validated twice) when we do user registration with some of our features turned on (i.e. that user registration goes through a different controller to manage having any part or field attached to the User ContentItem).

My current solution would be to "store" the result from google (using the data that we sent to their api as key) within the lifetime of the driver (i.e. a signle request). If the results are there already, the driver would not need to call the recaptcha endpoint again.

This approach seems to work for the use case I am testing it on, but I am not sure whether there are weird cases of situations where that behaviour is not wanted.
